File: DsAndAlgo/CPP/src/Practice/MaximumProduct.cpp
//
// Created by rk015348 on 17/05/25.
//

#include "../../include/practice/MaximumProduct.h"

#include <iostream>



std::vector<int> MaximumProduct::SplitNumberInDigits(int n) {
    if (n < 0) {
        throw std::invalid_argument("Negative number is not allowed");
    }
    if (n < 10) {
        return std::vector<int>{n};
    }
    std::vector<int> digits;
    while (n > 0) {
        digits.push_back(n % 10);
        n = n / 10;
    }
    return digits;
}


std::vector<int> MaximumProduct::RadixSort(std::vector<int> digits) {
    static const int RADIX = 10;
    std::vector<int> arr(RADIX, 0);
    std::vector<int> aux(digits.size(), 0);

    for (int digit : digits) {
        arr[digit]++;
    }
    for (int i = 1; i < RADIX; i++) {
        arr[i] += arr[i - 1];
    }

    for (size_t i = digits.size() - 1; i >= 0; i--) {
        aux[--arr[digits[i]]] = digits[i];
    }

    for (size_t i = 0; i < digits.size(); i++) {
        digits[i] = aux[i];
    }

    return digits;
}


int MaximumProduct::MaxProduct(int n) {
    try {
        std::vector<int> digits = SplitNumberInDigits(n);
        digits = RadixSort(digits);
        if (digits.size() < 2) {
            return 0;
        }
        int product = digits[digits.size() - 1] * digits[digits.size() - 2];
        return product;
    } catch (std::invalid_argument &e) {
        throw e;
    }
}


int MaximumProduct::MaxProductOptimal(int n) {
    int d1, d2, temp;
    d1 = n % 10;
    n /= 10;
    d2 = n % 10;
    n /= 10;
    while (n > 0) {
        temp = n % 10;
        if ((temp > d1) && (d1 < d2)) {
            d1 = temp;
            n = n/10;
            continue;
        }
        if (temp > d2) {
            d2 = temp;
        }
        n = n / 10;
    }
    return d1 * d2;
}

File: DsAndAlgo/Java/Coding/src/main/java/org/practice/MaximumProduct.java
package org.practice;


import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Given an integer, return the maximum product of any two digits in n.
 * example:
 *  31 -> 3
 *  124 -> 8 Max(1*2, 1*4, 2*4)
 */

public class MaximumProduct {

    /**
     * Split the number into digits and return the list of the digit.
     * Time Complexity => O(n)
     * Space Complexity => O(n)
     * @param n, number
     * @return List<Integer>, List containing digits
     * @throws IllegalArgumentException
     */
    List<Integer> splitNumberInDigits(int n) throws IllegalArgumentException {
        if (n < 0) {
            throw new IllegalArgumentException(String.format("Number %d is negative", n));
        }
        List<Integer> digits = new ArrayList<>();
        if (n < 10) {
            digits.add(n);
            return digits;
        }
        while (n > 0) {
            digits.add(n % 10);
            n /= 10;
        }
        //Collections.reverse(digits);
        return digits;
    }

    /**
     * Implementation of Radix Sort
     * Time Complexity -> O(n)
     * Space Complexity -> O(n)
     * @param digits, List of digits
     * @return List<Integer>,  List of sorted digits
     */
    private List<Integer> radixSort(List<Integer> digits) {
        Integer[] arr = new Integer[10];
        Arrays.fill(arr, 0);
        Integer[] aux = new Integer[digits.size()];
        for (Integer integer : digits) {
            arr[integer]++;
        }

        for (int i = 1; i < arr.length; i++) {
            arr[i] = arr[i] + arr[i - 1];
        }

        for (int i = digits.size() - 1; i >= 0; i--) {
            int digit = digits.get(i);
            aux[arr[digit] -1] = digit;
            arr[digit]--;
        }

        for (int i = 0; i < digits.size(); i++) {
            digits.set(i, aux[i]);
        }
        return digits;
    }

    /**
     * Method calculate maximum product of two digits from the number.
     * Time -> O(n) + O(n) + O(1) => O(n)
     * Space -> O(n)
     * @param n, number
     * @return integer, maximum product
     * @throws IllegalArgumentException
     */
    int maxProduct(int n) throws IllegalArgumentException {
        try {
            List<Integer> digits = splitNumberInDigits(n);
            if (digits.size() < 2) {
                return 0;
            }
            digits = radixSort(digits);
            return (digits.get(digits.size() - 1) * digits.get(digits.size() - 2));
        } catch (IllegalArgumentException e) {
            //e.printStackTrace();
            throw e;
        }
    }

    /** Optimal solution */
    int maxProduct2(int n) {
        int d1 = 0, d2 = 0, max = 0;
        d1 = n % 10;
        n =  n/10;
        d2 = n % 10;
        n = n/10;
        int temp;
        while(n > 0) {
            temp = n %10;
            if ((temp > d1) && (d1 < d2)) {
                d1 = temp;
                n = n/10;
                continue;
            }

            if (temp > d2) {
                d2 = temp;
            }
            n = n/10;
        }
        return d1 * d2;
    }
}
